### PR TITLE
Disable middlebox for TLS 1.3

### PR DESCRIPTION
**Pull request by technoLord**
Work done during IETF 103 Hackathon.

### CODE REVIEW

File: docs/cmdline-opts/ssl-disable-tls13-middlebox.d [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Long: ssl-disable-tls13-middlebox
+Help: Disable middlebox option for tls 1.3 
+Added: 7.25.0
+---
+This option tells curl to not use middlebox for tls 1.3.
```
Comment (danielgustafsson) [inline docs/cmdline-opts/ssl-disable-tls13-middlebox.d]: This should mention that it only applies to the OpenSSL backend.

File: lib/vtls/openssl.c [hunk 2]
```diff
@@ -2396,6 +2396,8 @@ static CURLcode ossl_connect_step1(struct connectdata *conn, int sockindex)
     ctx_options |= SSL_OP_NO_TLSv1_2;
 #ifdef TLS1_3_VERSION
     ctx_options |= SSL_OP_NO_TLSv1_3;
+    if (SSL_SET_OPTION(disable_tls13_middlebox))
+    SSL_CTX_clear_options(BACKEND->ctx, SSL_OP_ENABLE_MIDDLEBOX_COMPAT);
```
Comment (danielgustafsson) [inline lib/vtls/openssl.c]: Seems to have incorrect indentation, you can test by running `make checksrc`

File: docs/cmdline-opts/ssl-disable-tls13-middlebox.d [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Long: ssl-disable-tls13-middlebox
+Help: Disable middlebox option for tls 1.3 
+Added: 7.25.0
```
Comment (bagder) [inline docs/cmdline-opts/ssl-disable-tls13-middlebox.d]: Ideally, this will go into 7.63.0...

File: include/curl/curl.h [hunk 4]
```diff
@@ -795,6 +795,9 @@ typedef enum {
    servers, a user can this way allow the vulnerability back. */
 #define CURLSSLOPT_ALLOW_BEAST (1<<0)
 
+/* - TLSv1.3 MIDDLEBOX mode */
+#define CURLSSLOPT_TLS13_MIDDLEBOX (1<<0)
```
Comment (bagder) [inline include/curl/curl.h]: Surely this shouldn't use the same bit as "allow beast" ?

File: docs/cmdline-opts/ssl-disable-tls13-middlebox.d (hunk 1 above)
Comment (technoLord) [reply docs/cmdline-opts/ssl-disable-tls13-middlebox.d]: Yes I agree.

File: lib/vtls/openssl.c (hunk 2 above)
Comment (technoLord) [reply lib/vtls/openssl.c]: Correct indentation used.

File: docs/cmdline-opts/ssl-disable-tls13-middlebox.d (hunk 3 above)
Comment (technoLord) [reply docs/cmdline-opts/ssl-disable-tls13-middlebox.d]: Corrected.

File: include/curl/curl.h (hunk 4 above)
Comment (technoLord) [reply include/curl/curl.h]: New bit used for disable tls13 middlebox.

**Comment by mkauf** [comment]
The feature should probably be called "disable middlebox compatibility mode". The name "disable middlebox" is misleading.

File: src/tool_operate.c [hunk 5]
```diff
@@ -1473,6 +1473,7 @@ static CURLcode operate_do(struct GlobalConfig *global,
         /* new in 7.25.0 and 7.44.0 */
         {
           long mask = (config->ssl_allow_beast ? CURLSSLOPT_ALLOW_BEAST : 0) |
+		      (config->ssl_disable_tls13_middlebox ? CURLSSLOPT_TLS13_MIDDLEBOX: 0) |
```
Comment (vszakats) [inline src/tool_operate.c]: Indention is off by two spaces to the left here.

Comment (technoLord) [reply src/tool_operate.c]: Indentation fixed

File: lib/vtls/openssl.c (hunk 2 above)
Comment (vszakats) [reply lib/vtls/openssl.c]: Above indentation is still wrong and will cause warnings with certain C compilers.

File: src/tool_cfgable.h [hunk 6]
```diff
@@ -229,6 +229,8 @@ struct OperationConfig {
   bool xattr;               /* store metadata in extended attributes */
   long gssapi_delegation;
   bool ssl_allow_beast;     /* allow this SSL vulnerability */
+  bool ssl_disable_tls13_middlebox; /* disable tls 1.3 middlebox */
+  bool proxy_ssl_disable_tls13_middlebox; /* disable tls 1.3 middlebox for proxy */
```
Comment (danielgustafsson) [inline src/tool_cfgable.h]: This variable seems unused?

**Comment by danielgustafsson** [comment]
Ping @technoLord. Are you still interested in getting this PR in, or would you like some help with the last few things?

**Comment by technoLord** [comment]
Hello @danielgustafsson . I need guidance regarding the libcurl part.

**Comment by danielgustafsson** [comment]
> Hello @danielgustafsson . I need guidance regarding the libcurl part.

No worries, I've cleaned up a version of this patch with the libcurl parts fixed and will post a version for review soon.

**Comment by danielgustafsson** [comment]
Closing this as #3420, which extended this one, was closed.

**Pull request closed without merging**